source/agora/test/EnrollmentManager.d

```diff
@@ -221,3 +221,142 @@ unittest
+            if (data.tx_set.length == 0)
+                return;  // not ready yet
+
+            // Add an enrollment that is already an validator
+            data.enrolls ~= cast(Enrollment)BadNominator.dup_en;
+
+            // note: we are not passing the previous tx set as we don't really
+            // need it at this point (might later be necessary for chain upgrades)
+            auto slot_idx = this.ledger.getBlockHeight() + 1;
+            this.nominate(slot_idx, data);
+        }
+    }
+
+    static class MisbehavingValidator : TestValidatorNode
+    {
+        /// Ctor
+        public this (Config config, Registry* reg, immutable(Block)[] blocks)
+        {
+            super(config, reg, blocks);
+        }
+
+        ///
+        protected override Nominator getNominator (
+            NetworkManager network, KeyPair key_pair, Ledger ledger,
+            TaskManager taskman)
+        {
+            return new BadNominator(network, key_pair, ledger, taskman);
+        }
+    }
+
+    static class BadAPIManager : TestAPIManager
+    {
+        ///
+        public this (immutable(Block)[] blocks)
+        {
+            super(blocks);
+        }
+
+        /// see base class
+        public override void createNewNode (Config conf)
+        {
+            if (this.nodes.length == 1)
+            {
+                auto api = RemoteAPI!TestAPI.spawn!MisbehavingValidator(
+                    conf, &this.reg, this.blocks, conf.node.timeout);
+                this.reg.register(conf.node.address, api.tid());
+                this.nodes ~= NodePair(conf.node.address, api);
+            }
+            else
+            {
+                super.createNewNode(conf);
+            }
+        }
+    }
+
+    auto network = makeTestNetwork!BadAPIManager(TestConf.init);
+    scope(exit) network.shutdown();
+    scope(failure) network.printLogs();
+    network.start();
+    network.waitForDiscovery();
+    auto nodes = network.clients;
+    auto validator = nodes[0];
+    auto bad_validator = nodes[1];
+
+    // Sanity check: Check if genesis block has enrollments
+    const b0 = validator.getBlocksFrom(0, 2)[0];
+    assert(b0.header.enrollments.length >= 1);
+
+    // Set an duplicated enrollment to `BadNominator`
+    BadNominator.dup_en = cast(shared)b0.header.enrollments[0];
+
+    // Make a bloack using Genesis
+    Transaction[] txs;
+    txs = genesisSpendable().take(Block.TxsInBlock).enumerate()
+        .map!(en => en.value.refund(WK.Keys.A.address).sign())
+        .array();
+    txs.each!(tx => validator.putTransaction(tx));
+
+    // Ensure everyone is at the same level
+    ensureConsistency(nodes, 1);
+
+    // Make a block in the `bad_validator`
+    txs = txs
+        .map!(ptx => TxBuilder(ptx).refund(WK.Keys.B.address).sign())
+        .array();
+    txs.each!(tx => bad_validator.putTransaction(tx));
+
+    // Ensure the block is not accepted
+    ensureConsistency(nodes, 1);
```

Comment:
If you change this to:

```suggestion
    ensureConsistency(nodes, 2);
```

The tests will still pass. It's because the tests are not doing what you thought they were doing. :)

Reply:
^ That's right. I tested it the same way, but I still pass the test.

Reply:
Remember that when you send a transaction to a node, it is gossiped to all other nodes. All of the Validator nodes will nominate their own value. 1 of these nodes here will nominate a bad value (duplicate enrollment), however it doesn't matter because the rest of the nodes will reach consensus on the other nominations which are valid.

Reply:
There's a simpler test you can do:

- Spawn all BadValidator nodes (all 4 of them). Then send 8 transactions. Then do a thread.sleep call (to wait for propagation). Then check the block height of all nodes. Is it still height 0? That means nomination failed because the enrollment was rejected.

Actually it is very difficult to write correct networking tests right now because we don't get any information from the nodes - they don't tell us why the block height hasn't changed (was it a bad transaction? bad enrollment? no consensus reached?).

Reply:
I'm adding another validator called `OneTimeValidator` that is nominating only one time. Thank you.